### PR TITLE
rgw_sal_motr:[CORTX-27698] Checksum mismatch for GET API

### PR DESCRIPTION
**Pull request by dharmendra-jyani**
Signed-off-by: dharmendra-jyani <dharmendra.jyani@seagate.com>

### Problem Statement
Get object for s3cmd is giving checksum mismatch warning during getting the data from s3.

[root@ssc-vm-g3-rhev4-2880 ~]# s3cmd --no-ssl get s3://testbucket3/vmlinux vmlinux
download: 's3://testbucket3/vmlinux' -> 'vmlinux' [1 of 1]
10210440 of 10210440 100% in 0s 134.52 MB/s done
WARNING: MD5 signatures do not match: computed=92326b2cc6b6898b4c3f238dff8c0b10, received=f9c5cc6840ab0093475b3b0699f4e58b

### Fix
Bl is initialized in every iteration. So the new data is stored from the 0th index every time in bl. But in code we are passing 'off' to handle_data function. This is result in writing from different index and result in checksum mismatch, we have to pass 0 in place of off in handle_data function
 
### Parent Story
[https://jts.seagate.com/browse/CORTX-27698](url)
Checksum mismatch for GET API

### Subtask Covered
[https://jts.seagate.com/browse/CORTX-28762](url)
Reproduce issue of checksum mismatch error

[https://jts.seagate.com/browse/CORTX-28763](url)
Find RCA of issue

[https://jts.seagate.com/browse/CORTX-28764](url)
Implement fix for the issue

### Testing done

1. Object size 10 MB
[root@ssc-vm-g4-rhev4-0761 ~]# s3cmd --no-ssl get s3://testbucket3/test11 test11
download: 's3://testbucket3/test11' -> 'test11'  [1 of 1]
 10210455 of 10210455   100% in    0s    47.47 MB/s  done
[root@ssc-vm-g4-rhev4-0761 ~]#

2.  Object size 16 MB 
[root@ssc-vm-g4-rhev4-0761 ~]# s3cmd --no-ssl get s3://testbucket3/test99 test99
download: 's3://testbucket3/test99' -> 'test99'  [1 of 1]
 16777216 of 16777216   100% in    0s    96.83 MB/s  done

3. Object size 100 MB
[root@ssc-vm-g4-rhev4-0761 ~]# s3cmd --no-ssl get s3://testbucket3/test111 test111
download: 's3://testbucket3/test111' -> 'test111'  [1 of 1]
 104857600 of 104857600   100% in    0s   111.99 MB/s  done
[root@ssc-vm-g4-rhev4-0761 ~]#

4. Object size 130 MB
[root@ssc-vm-g4-rhev4-0761 ~]# s3cmd --no-ssl get s3://testbucket3/test112 test112
download: 's3://testbucket3/test112' -> 'test112'  [1 of 1]
 136314880 of 136314880   100% in    1s   117.55 MB/s  done
[root@ssc-vm-g4-rhev4-0761 ~]#
 

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [x] Includes bug reproducer
  - [ ] No tests

### CODE REVIEW

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -1825,10 +1825,13 @@ int MotrObject::read_mobj(const DoutPrefixProvider* dpp, int64_t off, int64_t en
...
-  actual = bs;
-  for (; off < end; off += actual) {
-    if (end - off < bs)
-        actual = end - off;
+  left = end - off;
+  for (; left > 0; off += actual) {
+    if (left < bs)
+      bs = this->get_optimal_bs(left);
+    actual = bs;
+    if (left < bs)
```
Comment (sachinpunadikar) [inline src/rgw/rgw_sal_motr.cc:1833]: Add comment stating that bs may remain same, and hence this additional check

File: src/rgw/rgw_sal_motr.cc [hunk 2]
```diff
@@ -1838,23 +1841,27 @@ int MotrObject::read_mobj(const DoutPrefixProvider* dpp, int64_t off, int64_t en
...
-    if (rc != 0)
+    if (rc != 0) {
+      ldpp_dout(dpp, 0) << __func__ << ": read failed, m0_op_wait rc=" << rc << dendl;
       goto out;
-
+    }
     // Call `cb` to process returned data.
     ldpp_dout(dpp, 20) << "MotrObject::read_mobj(): call cb to process data" << dendl;
-    cb->handle_data(bl, off, actual);
+    cb->handle_data(bl, 0, actual);
```
Comment (siningwuseagate) [inline src/rgw/rgw_sal_motr.cc:1864]: For byte-range GET, the offset `0` is not the correct offset in the bufferlist.
You can leave the test and fix of byte-range GET in a new PR.

File: src/rgw/rgw_sal_motr.cc (hunk 1 above)
Comment (andriytk) [reply src/rgw/rgw_sal_motr.cc:1834]: Sorry, I cannot get what you mean. `bs` can change - and that's why the additional check is needed.

**Comment by andriytk** [comment]
Please, remove all unrelated commits from this PR.

File: src/cephadm/cephadm [hunk 3]
```diff
@@ -3781,14 +3781,20 @@ class CephadmAgent():
         # Create the required config files in the daemons dir, with restricted permissions
         for filename in config:
             if filename in self.required_files:
-                with open(os.path.join(self.daemon_dir, filename), 'w') as f:
+                with open(os.open(os.path.join(self.daemon_dir, filename), os.O_CREAT | os.O_WRONLY, 0o600), 'w') as f:
```
Comment (andriytk) [inline src/cephadm/cephadm:3784]: Why are these changes here?

File: src/rgw/rgw_sal_motr.cc (hunk 2 above)
Comment (andriytk) [reply src/rgw/rgw_sal_motr.cc:1865]: Shouldn't it have the original value of `off` here which was passed to `MotrObject::read_mobj()`?

File: src/rgw/rgw_sal_motr.cc (hunk 1 above)
Comment (sjain09) [reply src/rgw/rgw_sal_motr.cc:1838]: This check is not for "bs value might not change" . This check is for if the new bs value is still more than `left` , in that case `actual` should be equal to the `left`. 

This comment is not appropriate and should be removed.

File: src/rgw/rgw_sal_motr.cc (hunk 2 above)
Comment (sjain09) [reply src/rgw/rgw_sal_motr.cc:1869]: That is the point. We are always writing data received from the motr into `bl` starting from 0th index in every iteration. We can not read from any other offset. That data will discarded at the end of every iteration. That will in-turn make sure that we are not accumulating too much data in `bl`.

Comment (andriytk) [reply src/rgw/rgw_sal_motr.cc:1869]: Sorry, I don't understand. This is a reading function, not writing. Why we cannot read from any other offset than zero?

Comment (andriytk) [reply src/rgw/rgw_sal_motr.cc:1869]: Sining explained:
> 0 is the offset local to the bl , not the offset in the whole object

**Review by bkirunge7 — APPROVED** [review]

**Review by andriytk — APPROVED** [review]

**Comment by dharmendra-jyani** [comment]
For Byte-range issue new ticket has been raised.
https://jts.seagate.com/browse/CORTX-29171

**Comment by andriytk** [comment]
```
commit 2e16031aa4ca2eb43b3ba2e9bd720d93477001c5 (HEAD -> main, sgt/main)
Author: dharmendra-jyani <91260838+dharmendra-jyani@users.noreply.github.com>
Date:   Fri Mar 4 13:07:44 2022 +0530

    rgw_sal_motr:[CORTX-27698] Checksum mismatch for GET API (#111)
```

Email address is wrong for `Author`.
